File: rust/var-core/src/cell_diff.rs
//! Table row/cell comparison — port of `cell-diff.ts` / `CellDiff.java`.
//! `Object` + `instanceof Map`/`List` duck-typing becomes matching on [`Value`].

use crate::ast::Table;
use crate::error::StepError;
use crate::span::Span;
use crate::value::Value;

/// The verdict for one checked column: expected vs actual, plus raw values and
/// whether a parameter-type `format` produced `actual` (inline-parameter path).
#[derive(Clone, Debug, PartialEq)]
pub struct CellDiff {
    pub column: String,
    pub span: Span,
    pub expected: String,
    pub actual: String,
    pub ok: bool,
    pub expected_value: Option<Value>,
    pub actual_value: Option<Value>,
    pub formatted: bool,
}

impl CellDiff {
    /// The five-component form (row/table paths): raw values `None`, not formatted.
    pub fn new(
        column: impl Into<String>,
        span: Span,
        expected: impl Into<String>,
        actual: impl Into<String>,
        ok: bool,
    ) -> CellDiff {
        CellDiff {
            column: column.into(),
            span,
            expected: expected.into(),
            actual: actual.into(),
            ok,
            expected_value: None,
            actual_value: None,
            formatted: false,
        }
    }
}

/// One checked column of one header-bound row.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RowCheck {
    pub column: String,
    pub value: String,
    pub span: Span,
}

impl RowCheck {
    pub fn new(column: impl Into<String>, value: impl Into<String>, span: Span) -> RowCheck {
        RowCheck {
            column: column.into(),
            value: value.into(),
            span,
        }
    }
}

/// Display rules 2–4 of the mismatch-rendering chain: a string as-is, anything
/// else a best-effort stringification.
pub fn render_cell_value(value: &Value) -> String {
    match value {
        Value::String(s) => s.clone(),
        Value::Int(i) => i.to_string(),
        Value::Float(d) => format!("{d}"),
        Value::Bool(b) => b.to_string(),
        Value::Null => "null".to_string(),
        // Port-native fallback (deliberately outside conformance): a bundle that
        // pins an object-valued actual must give the parameter type a `format`.
        Value::List(_) | Value::Map(_) => format!("{value:?}"),
    }
}

/// Compares a row step's returned map against the row's cells. Only columns
/// present on `returned` are checked; a non-map/`None` return checks nothing.
pub fn compare_row(returned: Option<&Value>, checks: &[RowCheck]) -> Vec<CellDiff> {
    let Some(Value::Map(obj)) = returned else {
        return Vec::new();
    };
    let mut diffs = Vec::new();
    for check in checks {
        let Some(value) = obj.get(&check.column) else {
            continue;
        };
        let actual = render_cell_value(value);
        let ok = actual == check.value;
        diffs.push(CellDiff::new(
            check.column.clone(),
            check.span,
            check.value.clone(),
            actual,
            ok,
        ));
    }
    diffs
}

/// Compares a whole-table step's returned table against the input table. `None`
/// checks nothing; type/shape problems return [`StepError::ReturnShape`].
pub fn compare_table(returned: Option<&Value>, input: &Table) -> Result<Vec<CellDiff>, StepError> {
    let rows = match returned {
        None => return Ok(Vec::new()),
        Some(Value::List(rows)) => rows,
        Some(other) => {
            return Err(StepError::ReturnShape(format!(
                "expected a table (array of rows), got {}",
                other.type_name()
            )));
        }
    };
    let columns = &input.header.cells;
    let data_rows = &input.rows;
    if rows.len() != data_rows.len() {
        return Err(StepError::ReturnShape(format!(
            "expected {} row(s), got {}",
            data_rows.len(),
            rows.len()
        )));
    }
    let all_arrays = rows.iter().all(|r| matches!(r, Value::List(_)));
    let all_records = rows.iter().all(|r| matches!(r, Value::Map(_)));
    if !all_arrays && !all_records {
        return Err(StepError::ReturnShape(
            "table rows must be all arrays or all objects".to_string(),
        ));
    }

    let mut diffs = Vec::new();
    for (i, (data_row, ret)) in data_rows.iter().zip(rows).enumerate() {
        if all_arrays {
            if let Value::List(cells) = ret {
                if cells.len() != columns.len() {
                    return Err(StepError::ReturnShape(format!(
                        "row {}: expected {} column(s), got {}",
                        i,
                        columns.len(),
                        cells.len()
                    )));
                }
            }
        }
        for (j, column) in columns.iter().enumerate() {
            let actual_value: &Value = if all_arrays {
                let Value::List(cells) = ret else {
                    unreachable!()
                };
                &cells[j]
            } else {
                let Value::Map(rec) = ret else { unreachable!() };
                match rec.get(column) {
                    Some(v) => v,
                    None => {
                        return Err(StepError::ReturnShape(format!(
                            "row {i}: missing column \"{column}\""
                        )));
                    }
                }
            };
            let expected = data_row.cells.get(j).map_or("", |c| c.as_str());
            let actual = render_cell_value(actual_value);
            let span = data_row.cell_spans.get(j).copied().unwrap_or(data_row.span);
            let ok = actual == expected;
            diffs.push(CellDiff::new(column.clone(), span, expected, actual, ok));
        }
    }
    Ok(diffs)
}

File: go/core/cell_diff.go
package core

import (
	"fmt"
	"strconv"
)

// Table row/cell comparison — port of cell-diff.ts / cell_diff.rs.

// CellDiff is the verdict for one comparison of one CELL — the atomic value a
// sensor checks against the document (a table cell, a header-bound row's cell,
// or a value captured from a paragraph). Expected vs actual, plus raw
// values and whether a parameter-type format produced actual.
type CellDiff struct {
	Column        string
	Span          Span
	Expected      string
	Actual        string
	Ok            bool
	ExpectedValue *Value
	ActualValue   *Value
	Formatted     bool
}

// newCellDiff is the five-component form (row/table paths): raw values nil, not
// formatted.
func newCellDiff(column string, span Span, expected, actual string, ok bool) CellDiff {
	return CellDiff{Column: column, Span: span, Expected: expected, Actual: actual, Ok: ok}
}

// RowCheck is one checked column of one header-bound row.
type RowCheck struct {
	Column string
	Value  string
	Span   Span
}

// NewRowCheck builds a RowCheck.
func NewRowCheck(column, value string, span Span) RowCheck {
	return RowCheck{Column: column, Value: value, Span: span}
}

// renderCellValue is display rules 2–4 of the mismatch-rendering chain: a string
// as-is, anything else a best-effort stringification.
func renderCellValue(value Value) string {
	switch value.Kind {
	case KindString:
		return value.Str
	case KindInt:
		return strconv.FormatInt(value.Int, 10)
	case KindFloat:
		return formatFloat(value.Float)
	case KindBool:
		if value.Bool {
			return "true"
		}
		return "false"
	case KindNull:
		return "null"
	default:
		// Port-native fallback (deliberately outside conformance): a bundle that
		// pins an object-valued actual must give the parameter type a format.
		return fmt.Sprintf("%v", value)
	}
}

// formatFloat renders a float the way the reference's `format!("{d}")` does — an
// integral value keeps no decimals, else shortest round-trip.
func formatFloat(d float64) string {
	if d == float64(int64(d)) {
		return strconv.FormatInt(int64(d), 10)
	}
	return strconv.FormatFloat(d, 'g', -1, 64)
}

// compareRow compares a row step's returned map against the row's cells. Only
// columns present on returned are checked; a non-map/nil return checks nothing.
func compareRow(returned *Value, checks []RowCheck) []CellDiff {
	if returned == nil || returned.Kind != KindMap {
		return nil
	}
	obj := returned.Map
	var diffs []CellDiff
	for _, check := range checks {
		value, ok := obj[check.Column]
		if !ok {
			continue
		}
		actual := renderCellValue(value)
		diffs = append(diffs, newCellDiff(check.Column, check.Span, check.Value, actual, actual == check.Value))
	}
	return diffs
}

// compareTable compares a whole-table step's returned table against the input
// table. nil checks nothing; type/shape problems return a SEReturnShape error.
func compareTable(returned *Value, input Table) ([]CellDiff, *StepError) {
	if returned == nil {
		return nil, nil
	}
	if returned.Kind != KindList {
		err := returnShapeError(fmt.Sprintf("expected a table (array of rows), got %s", returned.TypeName()))
		return nil, &err
	}
	rows := returned.List
	columns := input.Header.Cells
	dataRows := input.Rows
	if len(rows) != len(dataRows) {
		err := returnShapeError(fmt.Sprintf("expected %d row(s), got %d", len(dataRows), len(rows)))
		return nil, &err
	}
	allArrays := true
	allRecords := true
	for _, r := range rows {
		if r.Kind != KindList {
			allArrays = false
		}
		if r.Kind != KindMap {
			allRecords = false
		}
	}
	if !allArrays && !allRecords {
		err := returnShapeError("table rows must be all arrays or all objects")
		return nil, &err
	}

	var diffs []CellDiff
	for i := range dataRows {
		dataRow := dataRows[i]
		ret := rows[i]
		if allArrays {
			if len(ret.List) != len(columns) {
				err := returnShapeError(fmt.Sprintf("row %d: expected %d column(s), got %d", i, len(columns), len(ret.List)))
				return nil, &err
			}
		}
		for j, column := range columns {
			var actualValue Value
			if allArrays {
				actualValue = ret.List[j]
			} else {
				v, ok := ret.Map[column]
				if !ok {
					err := returnShapeError(fmt.Sprintf("row %d: missing column %q", i, column))
					return nil, &err
				}
				actualValue = v
			}
			expected := ""
			if j < len(dataRow.Cells) {
				expected = dataRow.Cells[j]
			}
			actual := renderCellValue(actualValue)
			span := dataRow.Span
			if j < len(dataRow.CellSpans) {
				span = dataRow.CellSpans[j]
			}
			diffs = append(diffs, newCellDiff(column, span, expected, actual, actual == expected))
		}
	}
	return diffs, nil
}
